File: src/main/java/decorator/decorator/TenOrderDiscount.java
package decorator.decorator;

import decorator.Discount;

public class TenOrderDiscount implements Discount {
    public static final double RATE = 0.02;
    public static final String HISTORY = "10번 주문 할인";
    private final Discount discount;

    public TenOrderDiscount(Discount discount) {
        this.discount = discount;
    }

    @Override
    public double amountAfterDiscount(double amount) {
        return discount.amountAfterDiscount(amount) * RATE;
    }

    @Override
    public String history() {
        return discount.history() + " " + HISTORY;
    }
}

File: kotlin/src/main/kotlin/decorator/decorator/TenOrderDiscount.kt
package decorator.decorator

import decorator.Discount

class TenOrderDiscount(private val discount: Discount): Discount {
    companion object {
        const val RATE = 0.02
        const val HISTORY = "10번 주문 할이"
    }

    override fun amountAfterDiscount(amount: Double): Double {
        return discount.amountAfterDiscount(amount) * RATE
    }

    override fun history(): String {
        return "${discount.history()} $HISTORY"
    }
}